src/protocols/secure_channel/PASESession.cpp

```diff
@@ -232,7 +232,18 @@ CHIP_ERROR PASESession::Pair(SessionManager & sessionManager, uint32_t peerSetUp
 exit:
     if (err != CHIP_NO_ERROR)
     {
+        // If a failure happens in PASESession::Init, we need to ensure that the passed exchange context is closed to prevent
+        // resource leaks.
+        exchangeCtxt->Close();
+
+        // If a failure happens in SendPBKDFParamRequest, we need to Discard the exchange (mExchangeCtxt) so that Clear() doesn't
+        // try closing it.  The exchange will handle that.
+        DiscardExchange();
         Clear();
+        ChipLogError(SecureChannel, "Failed during PASE session pairing request: %" CHIP_ERROR_FORMAT, err.Format());
+        MATTER_TRACE_COUNTER("PASEFail");
+        // Do this last in case the delegate frees us.
+        NotifySessionEstablishmentError(err);
```

Comment:
What's the API contract around this stuff?  It needs to be clearly documented.  For the exchange this is documented already, and we were just not fulfilling the API contract.  But for the delegate: do callers expect to get a call to the delegate if the `Pair()` call fails out?  I would not expect so, and certainly the one in CHIPDeviceController does not seem to expect it.  Concretely, in CHIPDeviceController.cpp the logic basically looks like this:

```
    err = device->GetPairing().Pair(*mSystemState->SessionMgr(), params.GetSetupPINCode(), GetLocalMRPConfig(), exchangeCtxt, this);
    if (err != CHIP_NO_ERROR)
    {
        ReleaseCommissioneeDevice(device);
    }
```

but now if we get `OnSessionEstablishmentError` before the `Pair()` call returns, that will call `RendezvousCleanup` which will do `ReleaseCommissioneeDevice()` on that same exact device pointer, which will do a double-free on the device.   Which might be a really bad idea.  So I think we should not be calling `NotifySessionEstablishmentError` here, and we should clearly document the API to explain when it _is_ called: Only after certain initial calls returned success.

Reply:
I agree, fixed.
where exactly do we document API contracts anyways, above the declaration of the method?